Add FOUCE utilities

claviska opened the pull request:
This is outside my split work, but I wanted to get it in to have an answer to what I expect to be one of the biggest problems autoloader/CDN users will have.

- Adds support for the `wa-cloak` class when using the autoloader. This eliminates FOUCE on the _initial load_ unless it takes more than two seconds to load.
- Adds the `wa-discovery-complete` class to the autoloader. This is used internally, but it can also be used by consumers to detect when autoloading is complete.
- Adds a utility for Turbo users to eliminate FOUCE when navigating between pages (I disabled SSR and tested this in our docs and it works great — try it out!)
- Adds this info to the appropriate sections of the docs
- Implements the `allDefined()` utility as mentioned in #486

---

- Related issue: #486
- Related support question: https://github.com/shoelace-style/webawesome-alpha/discussions/169

claviska committed:
add fouce utilities

claviska committed:
add comment

LeaVerou commented on docs/docs/installation.md:
```diff
@@ -37,8 +37,30 @@ This snippet includes three parts:
-While convenient, autoloading may lead to a [Flash of Undefined Custom Elements](https://www.abeautifulsite.net/posts/flash-of-undefined-custom-elements/). The linked article describes some ways to alleviate it.
+### Reducing FOUCE
+
+While convenient, autoloading can lead to a [Flash of Undefined Custom Elements](https://www.abeautifulsite.net/posts/flash-of-undefined-custom-elements/). To prevent this, you can add the `wa-reduce-fouce` class to any element on the page. Make sure to include the [FOUCE style utility](/docs/utilities/fouce/#opting-in) to get the corresponding CSS.
```
```suggestion
While convenient, autoloading can lead to a [Flash of Undefined Custom Elements](https://www.abeautifulsite.net/posts/flash-of-undefined-custom-elements/).
The [FOUCE style utility](/docs/utilities/fouce/#opting-in) (which is automatically applied if you use the Web Awesome utilities) takes care of hiding custom elements until they and their contents have been registered, up to a maximum of two seconds (configurable).

In many cases, this is not enough, and you may wish to hide a broader wrapper element or even the entire page until all WA elements within it have loaded. To do that, you can add the `wa-reduce-fouce` class to any element on the page or even apply it to the whole page by placing the class on the `<html>` or `<body>` element.
```

LeaVerou commented on src/styles/utilities/fouce.css:
```diff
@@ -18,3 +18,8 @@
+
+/* An opt in utility that hides any container with this class until the autoloader removes it */
+.wa-reduce-fouce {
+  opacity: 0;
```
This means that if the _autoloader itself_ fails to load, the page will be blank. We can piggyback on the CSS above, and then when the autoloader removes it, it's essentially just a performance optimization.

As a bonus, it gets the same customizable timeout.

LeaVerou replied:
Looks like I can't send an edit suggestion outside the diff. Basically this:

```diff
- :state(wa-defined):has(:not(:defined)) {
+ :state(wa-defined):has(:not(:defined)),
+ .wa-reduce-fouce:has(:not(:defined)) {
```

(actually, now that I think about it, this will stop applying once there are no undefined elements within, so do we even need the JS to remove the class?)

claviska replied:
updated in aa3cd97ddeb1073625965d32b0086fbd1da8e17b

LeaVerou commented:
Nice! I left some comments/suggestions. It seems to me that the CSS version _may_ actually reduce most of this JS, even for Turbo, do you want to give it a shot?

Not a blocker, but I’m wondering if we should avoid `fouce` in the name of this class, which is a bit esoteric. Some ideas:
- `wa-await-defined`
- `wa-cloak` (inspired from Vue, though also a bit cryptic)
- ?

claviska commented:
@LeaVerou Feel free to update the PR with whatever changes you see fit. I'm going to move on to my split work.

claviska committed:
Merge branch 'next' into fouce-class

claviska committed:
Update docs/docs/installation.md

Co-authored-by: Lea Verou <lea@verou.me>

claviska committed:
commit PR suggestion

claviska committed:
rename wa-reduce-fouce to wa-cloak

claviska committed:
Merge branch 'fouce-class' of https://github.com/shoelace-style/webawesome into fouce-class

claviska committed:
remove class as requested

claviska committed:
add cloak class

claviska committed:
wait a cycle

claviska committed:
move turbo to same file

claviska committed:
reduce fade

claviska committed:
disable SSR and add Turbo FOUCE helper

claviska commented:
I've renamed the class to `wa-cloak` and removed the CSS class, as requested.

[skipping long paragraph about why the themer picker's view transition tripped me up for over an hour]

New since the last review:

- Removed `.wa-cloak` CSS class
- Disabled SSR on Web Awesome
- Enabled Turbo FOUCE middleware

Let's see how the Vercel deployment looks.

claviska committed:
disable SSR

KonnorRogers commented on web-test-runner.config.js:
```diff
@@ -78,15 +76,8 @@ export default {
-        <script type="module">
-          ;(async () => {
-            await import('@lit-labs/ssr-client/lit-element-hydrate-support.js')
-            await Promise.allSettled(window.clientComponents.map(str => import(str)));
```
I believe this line needs to stay and is probably why its failing if i had to guess.

If you never want to run SSR, you would just move this back to the old autoloader.

claviska replied:
It doesn't seem to make a difference. I was just hoping to disable the SSR tests altogether since they've been failing and emailing us for weeks now.

LeaVerou commented on docs/docs/installation.md:
```diff
@@ -37,8 +37,33 @@ This snippet includes three parts:
 
-:::info
-While convenient, autoloading may lead to a [Flash of Undefined Custom Elements](https://www.abeautifulsite.net/posts/flash-of-undefined-custom-elements/). The linked article describes some ways to alleviate it.
+### Reducing FOUCE
```
Does it make sense to scatter the docs on FOUCE in two places? I would suggest moving most of this to the existing page on FOUCE utilities and just linking to it here.

claviska replied:
Feel free to make any changes you'd like to see to these docs.

KonnorRogers committed:
fix test suite

KonnorRogers committed:
workflow dispatch

LeaVerou commented:
Hi folks, what's the status on this? While not on the critical path for this split, it's really cool to be able to do navigations this fast without SSR, but if this is not ready for review, I could add `wa-cloak` in a separate PR, since that's a one line change (+ docs).

claviska commented:
@LeaVerou The tests are passing now so if you want docs changes, it looks like that would be the only thing left. Feel free to commit them directly to this branch if you want.

LeaVerou commented:
> @LeaVerou The tests are passing now so if you want docs changes, it looks like that would be the only thing left. Feel free to commit them directly to this branch if you want.

Happy to look into this once I finish a couple other things I'm doing. Are you sure you're okay with me committing directly? I don't mind opening another PR with this as target.

claviska committed:
Merge branch 'next' into fouce-class

claviska committed:
update fouce util

claviska committed:
no need to remove cloak class

claviska committed:
simplify fouce util

claviska committed:
add allDefined util

claviska committed:
update changelog

claviska commented:
This PR has been updated to include the `allDefined()` utility and simplify FOUCE handling/utils. I removed the customization and hard coded 2s for the timeout. We can add it back if enough users ask for it, but I suspect they won't and this seems like the perfect default. I also removed the example @LeaVerou added because the behaviors were identical after this update.

Ready to merge when approved.
